pkg/plugins/runtime/k8s/controllers/gatewayapi/http_route_controller.go

```diff
@@ -0,0 +1,184 @@
+	util_k8s "github.com/kumahq/kuma/pkg/util/k8s"
+)
+
+// HTTPRouteReconciler reconciles a GatewayAPI object into Kuma-native objects
+type HTTPRouteReconciler struct {
+	kube_client.Client
+	Reader kube_client.Reader
+	Log    logr.Logger
+
+	Scheme          *kube_runtime.Scheme
+	Converter       k8s_common.Converter
+	SystemNamespace string
+	ResourceManager manager.ResourceManager
+}
+
+// Reconcile handles transforming a gateway-api HTTPRoute into a Kuma
+// GatewayRoute and managing the status of the gateway-api objects.
+func (r *HTTPRouteReconciler) Reconcile(ctx context.Context, req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	httpRoute := &gatewayapi.HTTPRoute{}
+	if err := r.Get(ctx, req.NamespacedName, httpRoute); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			err = r.ResourceManager.Delete(ctx, core_mesh.NewGatewayRouteResource(), store.DeleteByKey(req.Name, req.Name))
+			if store.IsResourceNotFound(err) {
+				err = nil
+			}
+		}
+
+		return kube_ctrl.Result{}, err
+	}
+
+	// TODO check that gateways exist etc
+	// TODO set status on gapi resources
+
+	resource := core_mesh.NewGatewayRouteResource()
+
+	coreName := util_k8s.K8sNamespacedNameToCoreName(httpRoute.Name, httpRoute.Namespace)
+	mesh := k8s_util.MeshFor(httpRoute)
+
+	if err := r.ResourceManager.Get(ctx, resource, store.GetByKey(coreName, mesh)); err != nil {
+		if !store.IsResourceNotFound(err) {
+			return kube_ctrl.Result{}, err
+		}
+
+		resource.Spec, err = r.gapiToKumaRoute(httpRoute.Namespace, httpRoute)
+		if err != nil {
+			return kube_ctrl.Result{}, err
+		}
+
+		resource.SetMeta(&k8s.KubernetesMetaAdapter{ObjectMeta: metav1.ObjectMeta{
+			Name:      httpRoute.Name,
+			Namespace: httpRoute.Namespace,
+		}, Mesh: mesh})
+
+		err := r.ResourceManager.Create(ctx, resource, store.CreateByKey(coreName, mesh))
+
+		return kube_ctrl.Result{}, errors.Wrap(err, "could not create GatewayRoute resource")
+	}
+
+	var err error
+
+	resource.Spec, err = r.gapiToKumaRoute(httpRoute.Namespace, httpRoute)
+	if err != nil {
+		return kube_ctrl.Result{}, err
+	}
+
+	err = r.ResourceManager.Update(ctx, resource, store.ModifiedAt(core.Now()))
+
+	return kube_ctrl.Result{}, errors.Wrap(err, "could not update GatewayRoute resource")
+}
+
+func (r *HTTPRouteReconciler) gapiToKumaRoute(namespace string, route *gatewayapi.HTTPRoute) (*mesh_proto.GatewayRoute, error) {
+	var selectors []*mesh_proto.Selector
+
+	// Convert GAPI parent refs into Kuma tag matchers
+	for _, ref := range route.Spec.ParentRefs {
+		namespace := route.Namespace
+		if ns := ref.Namespace; ns != nil {
+			namespace = string(*ns)
+		}
+
+		match := map[string]string{
+			mesh_proto.ServiceTag: serviceTagForGateway(string(ref.Name), namespace),
+		}
+		if sn := ref.SectionName; sn != nil {
+			match[mesh_proto.ListenerTag] = string(*sn)
+		}
+		// TODO if sn == nil, what does it mean to reference the entire resource, all
+		// listeners? check this works
```

Comment:
Yeh, what is supposed to happen then is that the route applies to the whole Gateway, but you might still filter the matched routes by hostname (the listener hostname would have to match one of the route hostnames).

This is a bit complicated in the gateway plugin (see `RedistributeWildcardRoutes`. I'll file an issue because I'm not sure that this does the right thing if none of the route hostnames match on the listener.

Reply:
I checked and we have a test case for this

https://github.com/kumahq/kuma/blob/40260e27473a15504d690e2dcae008c020dde487/pkg/plugins/runtime/gateway/gateway_route_generator_test.go#L170-L173